server/src/main/java/io/crate/planner/optimizer/rule/ExtractConstantJoinConditionsToFilter.java

```diff
@@ -0,0 +1,155 @@
+ * <pre>
+ * Join[INNER | ((x = y) AND (y > 1))]
+ *   ├ Collect[doc.t1 | [x] | true]
+ *   └ Collect[doc.t2 | [y] | true]
+ * </pre>
+ * becomes:
+ * <pre>
+ * Filter[(y > 1)]
+ * └ Join[INNER | (x = y)]
+ *     ├ Collect[doc.t1 | [x] | true]
+ *     └ Collect[doc.t2 | [y] | true]
+ * </pre>
+ */
+public class ExtractConstantJoinConditionsToFilter implements Rule<JoinPlan> {
+
+    private static final Set<JoinType> SUPPORTED_JOIN_TYPES = EnumSet.of(INNER, LEFT, RIGHT, CROSS);
+    private final Pattern<JoinPlan> pattern;
+
+    public ExtractConstantJoinConditionsToFilter() {
+        this.pattern = typeOf(JoinPlan.class)
+            .with(join ->
+                SUPPORTED_JOIN_TYPES.contains(join.joinType()) &&
+                !join.joinConditionsExtractedToFilters() &&
+                join.joinCondition() != null &&
+                hasConstantJoinConditions(join.joinCondition())
+            );
+    }
+
+    private static boolean hasConstantJoinConditions(Symbol joinCondition) {
+        for (var condition : QuerySplitter.split(joinCondition).values()) {
+            if (numberOfRelationsUsed(condition) <= 1) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Pattern<JoinPlan> pattern() {
+        return pattern;
+    }
+
+    @Override
+    public LogicalPlan apply(JoinPlan joinPlan,
+                             Captures captures,
+                             PlanStats planStats,
+                             TransactionContext txnCtx,
+                             NodeContext nodeCtx,
+                             UnaryOperator<LogicalPlan> resolvePlan) {
+        var conditions = joinPlan.joinCondition();
+        var allConditions = QuerySplitter.split(conditions);
+        var constantConditionsForFilter = new LinkedHashSet<Symbol>(allConditions.size());
+        var remainingConditions = new LinkedHashSet<Symbol>(allConditions.size());
+        var lhsRelations = new HashSet<>(joinPlan.lhs().relationNames());
+        var rhsRelations = new HashSet<>(joinPlan.rhs().relationNames());
+        for (var condition : allConditions.entrySet()) {
+            Set<RelationName> relationsInCurrentConditions = RelationNameCollector.collect(condition.getValue());
+            if (relationsInCurrentConditions.size() == 1) {
+                // Constant join conditions can only pushed to the preserved side of a join, where no null values are created:
+                // - lhs/rhs sides of cross/inner joins
+                // - lhs of left join
+                // - rhs of right join
+                if (joinPlan.joinType().isOuter()) {
+                    var matchesLhs = Sets.intersection(lhsRelations, relationsInCurrentConditions);
+                    var matchesRhs = Sets.intersection(rhsRelations, relationsInCurrentConditions);
+                    if ((!matchesLhs.isEmpty() && joinPlan.joinType() == LEFT) ||
```

Comment:
I compared to postgres and it seems that this check should be inverted.
I still don't get it, it seems counter intuitive, but check it out:
psql:
```
matriv=> create table t1(a int);
CREATE TABLE
matriv=> create table t2(b int);
CREATE TABLE
matriv=> insert into t1 select * from generate_series(1, 10, 1);
INSERT 0 10
matriv=> insert into t2 select * from generate_series(6, 10, 1);
INSERT 0 5

matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
 a  | b  
----+----
  1 |   
  2 |   
  3 |   
  4 |   
  5 |   
  6 |  6
  7 |  7
  8 |  8
  9 |  9
 10 | 10
(10 rows)
matriv=> explain SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
                            QUERY PLAN                            
------------------------------------------------------------------
 Merge Left Join  (cost=359.57..941.28 rows=10838 width=8)
   Merge Cond: (t1.a = t2.b)
   Join Filter: (t1.a > 5)
   ->  Sort  (cost=179.78..186.16 rows=2550 width=4)
         Sort Key: t1.a
         ->  Seq Scan on t1  (cost=0.00..35.50 rows=2550 width=4)
   ->  Sort  (cost=179.78..186.16 rows=2550 width=4)
         Sort Key: t2.b
         ->  Seq Scan on t2  (cost=0.00..35.50 rows=2550 width=4)
(9 rows)

matriv=> SELECT * FROM t1 RIGHT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
 a  | b  
----+----
  6 |  6
  7 |  7
  8 |  8
  9 |  9
 10 | 10
(5 rows)
matriv=> explain SELECT * FROM t1 RIGHT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
                            QUERY PLAN                            
------------------------------------------------------------------
 Merge Right Join  (cost=263.02..429.84 rows=10838 width=8)
   Merge Cond: (t1.a = t2.b)
   ->  Sort  (cost=83.23..85.36 rows=850 width=4)
         Sort Key: t1.a
         ->  Seq Scan on t1  (cost=0.00..41.88 rows=850 width=4)
               Filter: (a > 5)
   ->  Sort  (cost=179.78..186.16 rows=2550 width=4)
         Sort Key: t2.b
         ->  Seq Scan on t2  (cost=0.00..35.50 rows=2550 width=4)
(9 rows)
```

crate (this PR):
```
cr> SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
+----+----+
|  a |  b |
+----+----+
|  7 |  7 |
|  8 |  8 |
|  9 |  9 |
| 10 | 10 |
|  6 |  6 |
+----+----+
SELECT 5 rows in set (1.223 sec)
cr> SELECT * FROM t1 RIGHT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
+----+----+
|  a |  b |
+----+----+
|  7 |  7 |
|  8 |  8 |
|  9 |  9 |
| 10 | 10 |
|  6 |  6 |
+----+----+
SELECT 5 rows in set (0.988 sec)
```


checkout how psql produces different result if the constant is moved from `ON` to `WHERE`:
```
matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 5 AND t1.a = t2.b;
 a  | b  
----+----
  1 |   
  2 |   
  3 |   
  4 |   
  5 |   
  6 |  6
  7 |  7
  8 |  8
  9 |  9
 10 | 10
(10 rows)
                                        ^
matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a = t2.b WHERE t1.a > 5;
 a  | b  
----+----
  6 |  6
  7 |  7
  8 |  8
  9 |  9
 10 | 10
(5 rows)

```

Can you please check some other database?
Can you also run the `crate-qa` sqllogictests against your branch to see if it breaks something?

Reply:
To demonstrate it better (filter is removing more rows than the matches between t1 and t2):
```
matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a > 8 AND t1.a = t2.b;
 a  | b  
----+----
  1 |   
  2 |   
  3 |   
  4 |   
  5 |   
  6 |   
  7 |   
  8 |   
  9 |  9
 10 | 10
(10 rows)

matriv=> SELECT * FROM t1 LEFT JOIN t2 ON t1.a = t2.b WHERE t1.a > 8;
 a  | b  
----+----
  9 |  9
 10 | 10
(2 rows)

```